pepper-apis/src/main/java/org/broadinstitute/ddp/cache/CacheService.java

```diff
@@ -0,0 +1,153 @@
+                if (mapper instanceof IdToCacheKeyCollectionMapper) {
+                    cacheNameToCacheKeyCollectionMapper.put(cacheName, (IdToCacheKeyCollectionMapper) mapper);
+                } else if (mapper instanceof IdToCacheKeyMapper) {
+                    cacheNameToCacheKeyMapper.put(cacheName, (IdToCacheKeyMapper) mapper);
+                } else {
+                    throw new DDPException("Unrecognized mapper type");
+                }
+                updateChangeTypeToCacheName(evictionModelChangeType, cacheName);
+                return cache;
+            }
+        } else {
+            return (Cache<K, V>) existingCache;
+        }
+    }
+
+    private <K, V> Cache<K, V> _getOrCreateCache(String cacheName, Duration entryDuration) {
+        Cache cache = cacheManager.getCache(cacheName);
+        if (cache == null || cache.isClosed()) {
+            var cacheConfig = new MutableConfiguration<K, V>()
+                    .setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(entryDuration))
+                    .setStatisticsEnabled(true);
+            cache = cacheManager.createCache(cacheName, cacheConfig);
+        }
+        return (Cache<K, V>) cache;
+    }
+
+    public void destroyCache(String name) {
+        cacheManager.destroyCache(name);
+    }
+
+    private void updateChangeTypeToCacheName(ModelChangeType evictionModelChangeType, String cacheName) {
+        Collection<String> existingCacheNames = modelChangeTypeToCacheName.get(evictionModelChangeType);
+        List<String> newListOfCacheNames;
+        if (existingCacheNames == null) {
+            newListOfCacheNames = List.of(cacheName);
+        } else {
+            newListOfCacheNames = new ArrayList<>(existingCacheNames);
+            newListOfCacheNames.add(cacheName);
+        }
+        modelChangeTypeToCacheName.put(evictionModelChangeType, newListOfCacheNames);
+    }
+
+    public void modelUpdated(ModelChangeType changeType, Handle handle, long id) {
+        findCacheNameByEventType(changeType).forEach(cacheName -> {
+            Cache cache = cacheManager.getCache(cacheName);
+            if (cache != null) {
+                IdToCacheKeyMapper<?> mapper = cacheNameToCacheKeyMapper.get(cacheName);
+                if (mapper != null) {
+                    Object cacheKey = cacheNameToCacheKeyMapper.get(cacheName).mapToKey(id, handle);
+                    if (cacheKey != null) {
+                        cache.remove(cacheKey);
+                    }
+                } else {
+                    IdToCacheKeyCollectionMapper<?> keyCollectionMapper = cacheNameToCacheKeyCollectionMapper.get(cacheName);
+
+                    if (keyCollectionMapper != null) {
+                        Set<?> keys = keyCollectionMapper.mapToKeys(id, handle);
+                        if (keys != null) {
+                            cache.removeAll(keys);
+                        }
+                    }
+                }
+            }
+        });
+    }
+
```

Comment:
Here we process notifications from (primarily) daos that they have modified (updated or inserted) data that would make the cached data no longer valid. Try to find the affected cached key entries by the change type and an id.
Currently we are only worrying about ModelChangeType.USER and we get the DB user id. So any changes to the user will trigger evicting data that has been cached about that user.